src/common/system_logger.h

```diff
@@ -0,0 +1,91 @@
+#ifndef GFS_COMMON_SYSTEM_LOGGER_H_
+#define GFS_COMMON_SYSTEM_LOGGER_H_
+
+#include <glog/logging.h>
+#include <glog/stl_logging.h>
+
+#include <string>
+
+// The system_logger wraps the logging library, which in this case is Google
+// glog. The reason for having a wrapper is to ensure that we can easily change
+// the underlying logging library if needed without having to change the
+// system_logger clients code. Also this helps us to control the functionalities
+// the underlying library that we want to expose.
+
+namespace gfs {
+
+namespace common {
+
+// Singleton class for managing for configuring how we log. The system logger
+// must be intialized in the application (ideally once) before using any of the
+// logging macros. Please see examples/system_logger_main.cc on how to use.
+// We currently only log to console.
+class SystemLogger {
+ public:
+  static SystemLogger& GetInstance() {
+    static SystemLogger instance;
+
+    return instance;
+  }
+
+  // To prevent copying the SystemLogger
+  SystemLogger(const SystemLogger&) = delete;
+
+  void operator=(const SystemLogger&) = delete;
+
+  // This initializes the logging framework and does the necessary
+  // configurations. Currently logging to console only. Must be called before
+  // using any of the logging macros. This isn't thread safe and must be done
+  // from a single thread.
+  void Initialize(const std::string& program_name);
+
+ private:
+  bool is_initialized_;
+
+  SystemLogger() : is_initialized_(false) {}
+};
+
+}  // namespace common
+}  // namespace gfs
+
+// The macros you should use for logging. only use the SYS_* macros. Please see
+// examples/system_logger_main.cc on how to use.
+//
+// SEVERITY LEVELS: You can specify one of the following severity levels (in
+// increasing order of severity): INFO, WARNING, ERROR, and FATAL. Logging a
+// FATAL message terminates the program (after the message is logged). Note that
+// messages of a given severity are logged not only in the logfile for that
+// severity, but also in all logfiles of lower severity. E.g., a message of
+// severity FATAL will be logged to the logfiles of severity FATAL, ERROR,
+// WARNING, and INFO.
+// Further reading about glog:
+// https://github.com/google/glog/blob/master/doc/glog.html
+
+// Log with the specified severity.
+#define SYS_LOG(severity) LOG(severity)
```

Comment:
FYI. Google Logging already defines thread safe logging macros of all varieties: 
https://hpc.nih.gov/development/glog.html
https://github.com/google/glog/blob/master/src/glog/logging.h.in

Can we simply use theirs

An example:
https://github.com/google/glog/blob/master/bazel/example/main.cc

Reply:
This is what I'm using here right. Please look at the src/common/system_logger.h (also added couple of comments there), essentially wrapping around the glog stuffs, but using the thread safe ones. The macro LOG(severity) is by glog and used for  LOG(INFO), LOG(WARNING) etc. Not sure what the actual concern here is?

Reply:
I think LOG is much cleaner and readable and I prefer it over SYS_LOG. If we change the underlying logging library, we just need to at that time define LOG to point to the new library macros. So I don't see any issue of using a simpler LOG which is more readable

Reply:
In addition to ease of changing the underlying framework. You wrap external framework to also control what functionalities in that framework you want to expose or not expose in your system. There are some logging macros in the lib that we shouldn't use and by wrapping it with ours, folks will only use those we expose. Also it's not only LOG MACROs we expose, there are macros like CHECK, (which I think it isn't well named) and wrapped it as ASSERT (since this is the well known term). Doing this also makes sure that we don't conform our code to some external style (Good or bad). If we use CHECK for example and change to a library that uses Assert (e.g. Log4cxx), then you are stuck. You have to remove all the CHECKs in your code base and change them to ASSERT or redefine CHECK as ASSERT (now you end up redefining). I know this is most likely not going to happen for this project because it's small and not a product. But It's hard for me not to follow good design principles just for the sake of the project being small. Plus most times following the right design choice doesn't usually take a lot of time.   But your point about LOG is valid, it's simpler. But I also don't think SYS_LOG is difficult to use. Will gladly take a better naming suggestion, as long as the underlying design principle here, which is having an abstraction layer is maintained. This helps to build a clean system that can easily evolve. Sorry for the long response. Nice discussion.

Reply:
I partially disagree. The way you claim as the "good design practice" is not the good design practice. A good design is a design that supports extensibility and backwards compatibility without unnecessary engineering overhead.

1. making a wrapper around another logging library doesn't prevent anyone from using other macros as they are still available; the way to prevent people from using it is to disallow it in code review and enforce through code review linter tools
2. we don't need to conform to external code style, but we should conform to good and generic code style. `LOG` is clearly way more extensible and generic and actually informative than our custom unnecessary wrapper `SYS_LOG`. However, you only need wrappers around non conventional macros such as defining  `ASSERT` to `CHECK`. But for all other perfectly good ones like `LOG`, we should keep them.
3. Anytime that you try to introduce new breaking API implementation changes, which happens quite often in industry, you (1) provide depreciation warnings for people to migrate, (2) introduce backwards compatible wrappers that uses old API convention but with new implementation; for example, if you are trying to deprecate `CHECK` that Log4cxx doesn't support, you simply define a new macro `CHECK` that uses Log4cxx's `ASSERT`; (3) a mass find and replace for codebase you control is also perfectly fine. 

This is the right way for a system to evolve, because if you claim that your initial wrapper that basically wraps around basically the same thing underneath is good for so called evolve, then any drastic breaking core implementation changes always have to (1) consider backwards compatibility, (2) has a very strong case for the drastic change; for example, a supreme benefit; (3) can provide backwards compatibility by overload existing conventions with new implementations, instead of having an unnecessary additional wrapper A over B, when you can always overload B's implementation.

Regardless, I strongly feel the need to use `LOG` convention not `SYS_LOG`, but I am fine with redefining `CHECK` to `ASSERT`

Reply:
You obviously missed my point, and discussions like this will be interesting if it doesn't become about "my way is better" and folks listen and understand what is being said. In the end of my previous message I stated "**But your point about LOG is valid, it's simpler**. But I also don't think SYS_LOG is difficult to use. **Will gladly take a better naming suggestion, as long as the underlying design principle here, which is having an abstraction layer is maintained**." And from what I get, you aren't saying the abstraction layer should be removed, you just want LOG* to be maintained. **Which I acknowledged as being valid. Will make the change for that.** Generally, discussions like this will go well if it is kept positive and civil. There is no value in a non positive conversation.